polytracker/cfg.py

```diff
@@ -0,0 +1,177 @@
+import math
+
+from typing import (
+    Any,
+    Callable,
+    Collection,
+    Dict,
+    FrozenSet,
+    Generic,
+    ItemsView,
+    Iterable,
+    KeysView,
+    List,
+    Optional,
+    Set,
+    TypeVar,
+)
+
+import graphviz
+import networkx as nx
+
+N = TypeVar("N")
+
+
+class DiGraph(nx.DiGraph, Generic[N]):
```

Comment:
Could you explain this python to me? So you defined type N which I guess is the "Node" type, and you are using it to be a generic DiGraph right

Reply:
This is solely for type hinting. It's basically the equivalent of this C++:
```c++
typedef <typename N>
class DiGraph : public nx::DiGraph { ... }
```
You can then use `N` within type hints through the rest of the class, like this:
https://github.com/trailofbits/polytracker/blob/e0af394394ea7c22c3e7b637230c3c7436c0529c/polytracker/cfg.py#L55-L56
And when you create an instance of that class, you can optionally provide a type for `N`, like I do here:
https://github.com/trailofbits/polytracker/blob/e0af394394ea7c22c3e7b637230c3c7436c0529c/polytracker/cfg.py#L157
So a `DiGraph` can have nodes of any type. But a `CFG`, which is a subclass of `DiGraph` specifically has nodes of type `FunctionInfo`. However, you could declare a DiGraph with _any_ type:
```python
graph: DiGraph[int] = DiGraph()
graph.add_node(1) # This is fine
graph.add_node('foo') # This will work but trigger a mypy/type hinting error
for root in graph.roots:
    bar: float = root # This will also cause a mypy/type hinting error
```